src/Microsoft.Diagnostics.DebugServices.Implementation/ModuleServiceFromDataReader.cs

```diff
@@ -0,0 +1,120 @@
+            {
+                _moduleService = moduleService;
+                _moduleInfo = moduleInfo;
+                ModuleIndex = moduleIndex;
+            }
+
+            #region IModule
+
+            public override int ModuleIndex { get; }
+
+            public override string FileName => _moduleInfo.FileName;
+
+            public override ulong ImageBase => _moduleInfo.ImageBase;
+
+            public override ulong ImageSize => (uint)_moduleInfo.IndexFileSize;
+
+            public override int IndexFileSize => _moduleInfo.IndexFileSize;
+
+            public override int IndexTimeStamp => _moduleInfo.IndexTimeStamp;
+
+            public override VersionInfo? Version
+            {
+                get 
+                {
+                    if (InitializeValue(Module.Flags.InitializeVersion))
+                    {
+                        if (_moduleInfo.Version != EmptyVersionInfo)
+                        {
+                            base.Version = _moduleInfo.Version;
+                        }
+                        else
+                        {
+                            if (_moduleService.Target.OperatingSystem != OSPlatform.Windows)
+                            {
+                                GetVersionFromVersionString();
+                            }
+                        }
+                    }
+                    return base.Version;
+                }
+            }
+
+            public override string VersionString
+            {
+                get
+                {
+                    if (InitializeValue(Module.Flags.InitializeProductVersion))
+                    {
+                        if (_moduleService.Target.OperatingSystem != OSPlatform.Windows && !IsPEImage)
+                        {
+                            _versionString = _moduleService.GetVersionString(ImageBase, ImageSize);
+                        }
+                    }
+                    return _versionString;
+                }
+            }
+
+            #endregion
+
+            protected override ModuleService ModuleService => _moduleService;
+        }
+
+        private readonly IDataReader _dataReader;
+
+        public ModuleServiceFromDataReader(ITarget target, IDataReader dataReader)
+            : base(target)
+        {
+            _dataReader = dataReader;
+        }
+
+        /// <summary>
+        /// Get/create the modules dictionary.
+        /// </summary>
+        protected override Dictionary<ulong, IModule> GetModulesInner()
+        {
+            var modules = new Dictionary<ulong, IModule>();
+            int moduleIndex = 0;
+
+            foreach (ModuleInfo moduleInfo in _dataReader.EnumerateModules().OrderBy((info) => info.ImageBase))
+            {
+                var module = new ModuleFromDataReader(this, moduleIndex, moduleInfo);
+                if (!modules.TryGetValue(moduleInfo.ImageBase, out IModule dup))
```

Comment:
If you expect the duplicates to be rare, maybe use `TryAdd`, and do a second read in the fail path.

Reply:
Dups are rare.

There isn't a Dictionary.TryAdd in netstandard2.0 as far as I can tell. Or are you saying to just use Add and catch the ArgumentException if there is a dup?

Reply:
I changed it to catch the exception.